File: aoj/intro/GRL/GRL_4_B.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int V, E;
vector<int> G[10010];
vector<int> ans;
int indeg[10010];
bool used[10010];

void bfs(int s) {
    queue<int> que;
    que.push(s);
    used[s] = true;
    while (!que.empty()) {
        int u = que.front(); que.pop();
        ans.push_back(u);
        for (auto v: G[u]) {
            indeg[v]--;
            if (indeg[v] == 0 && !used[v]) {
                used[v] = true;
                que.push(v);
            }
        }
    }
}

void tsort() {
    for (int u = 0; u < V; u++) {
        used[u] = false;
        indeg[u] = 0;
    }

    for (int u = 0; u < V; u++)
        for (auto v: G[u]) indeg[v]++;

    for (int u = 0; u < V; u++)
        if (indeg[u] == 0 && !used[u]) bfs(u);
}


int main(void) {
    cin >> V >> E;
    for (int i = 0; i < E; i++) {
        int s, t;
        cin >> s >> t;
        G[s].push_back(t);
    }

    tsort();
    for (auto i: ans) cout << i << endl;

    return 0;
}

File: aoj/intro/GRL/GRL_4_B.py
def tsort(G):
    def bfs(s):
        que = [s]
        used[s] = True
        while que:
            u = que.pop(0)
            res.append(u)

            for v in G[u]:
                indeg[v] -= 1
                if indeg[v] == 0 and not used[v]:
                    used[v] = True
                    que.append(v)

    indeg = [0] * V
    used = [False] * V
    res = []
    for u in xrange(V):
        for v in G[u]:
            indeg[v] += 1
    for u in xrange(V):
        if indeg[u] == 0 and not used[u]:
            bfs(u)
    return res


V, E = map(int, raw_input().split())
G = [[] for i in xrange(V)]
for i in xrange(E):
    s, t = map(int, raw_input().split())
    G[s].append(t)

ans = tsort(G)
for i in ans:
    print i
